aspnetcore/web-api/action-return-types.md

```diff
@@ -36,60 +36,99 @@ This document explains when it's most appropriate to use each return type.
 
 The simplest action returns a primitive or complex data type (for example, `string` or a custom object type). Consider the following action, which returns a collection of custom `Product` objects:
 
-[!code-csharp[](../web-api/action-return-types/samples/WebApiSample.Api.21/Controllers/ProductsController.cs?name=snippet_Get)]
+[!code-csharp[](../web-api/action-return-types/samples/2x/WebApiSample.Api.21/Controllers/ProductsController.cs?name=snippet_Get)]
 
 Without known conditions to safeguard against during action execution, returning a specific type could suffice. The preceding action accepts no parameters, so parameter constraints validation isn't needed.
 
-When known conditions need to be accounted for in an action, multiple return paths are introduced. In such a case, it's common to mix an [ActionResult](/dotnet/api/microsoft.aspnetcore.mvc.actionresult) return type with the primitive or complex return type. Either [IActionResult](#iactionresult-type) or [ActionResult\<T>](#actionresultt-type) are necessary to accommodate this type of action.
+When known conditions need to be accounted for in an action, multiple return paths are introduced. In such a case, it's common to mix an <xref:Microsoft.AspNetCore.Mvc.ActionResult> return type with the primitive or complex return type. Either [IActionResult](#iactionresult-type) or [ActionResult\<T>](#actionresultt-type) are necessary to accommodate this type of action.
+
+::: moniker range=">= aspnetcore-3.0"
+
+### Asynchronous stream
+
+Memory consumption on the server becomes a performance consideration when returning large data sets. Imagine an action that calls <xref:System.Linq.Enumerable.ToList*> on a LINQ query. The query returns an entire product catalog, consisting of hundreds of thousands of products. Each product record is stored in server memory. An asynchronous stream can relieve the burden on the server.
```

Comment:
This is not true sadly. We don't stream individual records into the response.

What this feature does is make using thing that return `IAsyncEnumerable<>` as efficient as `IEnumerable<>`.

Reply:
The problem that we solved:

- If you returned an `IAsyncEnumerable<>` in 2.2 we would iterated it **synchronously** - which results in blocking calls and potential thread-pool starvation. The work we did in 3.0 makes this naive code just as efficent as using `IEnumerable<>`.

We can't implement streaming for real until the JSON serializer has it.